### PR TITLE
Open data sets or directories via URL fragment

### PR DESCRIPTION
**Pull request by sk1p**
Example: http://localhost:9000/#action=open&path=/path/to/your/data/

The path can point to either a directory or a file, and will open the file browser or directly the open dialog, once LiberTEM is ready for loading data.

Part of #1085

## Contributor Checklist:

* [ ] I have added or updated my entry in [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [x] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [x] I have added/updated documentation for all user-facing changes
* [x] I have added/updated test cases
* [x] I have included the [rebuilt production build of the client](https://libertem.github.io/LiberTEM/contributing.html?#building-the-client) (only if changes were made to the GUI)

## Reviewer Checklist:

* [x] `/azp run libertem.libertem-data` passed
* [ ] No import of GPL code from MIT code

### CODE REVIEW

**Comment by sk1p** [comment]
Coverage _should_ be better in this PR, but isn't, probably because of https://github.com/LiberTEM/LiberTEM/issues/1472 :( :( - only 7 of the jobs managed to upload coverage...

**Comment by sk1p** [comment]
I've triggered a re-run of CI, in the hope we don't hit rate limits this time... *sigh* - wasting CI minutes because of a too-low rate limit shouldn't be the way things work.

**Review by matbryan52 — COMMENTED** [review]
Have tried to test / break the feature as best I can, combined with my knowledge of the TS of the client, and I think it's OK!

One observation on the behaviour: if for whatever reason the user's page is refreshed the system will re-parse the argument and create a second browser window (or open window), or error message if the arg is invalid. I can see pros and cons to this. We could cache the args we have parsed already and skip them on subsequent runs, but then the user doesn't get easy re-loading of data if they accidentally close the dataset. One downside of the current setup is they may see the same modal error messages multiple times.

I think a complementary PR would be to implement command-line arguments to `libertem-server` which allow pre-loading a cluster and a dataset / working directory (#1419). I will see if I can make this happen. One method I can think of would be to support an `action=cluster` or similar in the URL path, then adjust the auto-launch URL from the python CLI to fill these parameters.

File: src/libertem/io/fs.py [hunk 1]
```diff
@@ -49,6 +49,23 @@ def _get_alt_path(path):
     return cur_path
 
 
+def stat_path(path: str):
+    try:
+        return Path(path).resolve().stat()
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/io/fs.py:54]: True, that will also need to be changed in the fs listing code, I'll update it.

File: client/src/browser/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    const byParam = hash.split('&');
+
+    return Object.fromEntries(
+        byParam.map(paramStr => splitLikePython(paramStr, '=', 1))
+    ) as HashParameters; // yuck, fromEntries doesn't propagate the types from its argument
+}
+
+export const getUrlAction = (): URLAction => {
```
[the remark this answers is not in the document]
Comment (sk1p) [reply client/src/browser/helpers.ts:47]: Yes, I'll work on the error handling for this, good catch.

**Comment by sk1p** [comment]
> Have tried to test / break the feature as best I can, combined with my knowledge of the TS of the client, and I think it's OK!



> One observation on the behaviour: if for whatever reason the user's page is refreshed the system will re-parse the argument and create a second browser window (or open window), or error message if the arg is invalid. I can see pros and cons to this. We could cache the args we have parsed already and skip them on subsequent runs, but then the user doesn't get easy re-loading of data if they accidentally close the dataset. One downside of the current setup is they may see the same modal error messages multiple times.

Yeah, I decided on this behavior, to enable the user to repeat whatever action is given in the URL. Otherwise, the behaviour would be a bit non-transparent to the user, as they can't easily inspect or understand where this caching is happening. Also important for longer-running LiberTEM instances, for example as part of a service offering (think https://github.com/LiberTEM/LiberTEM-jupyter-proxy) - if you want to re-open a data set that was already opened at some point of the life time of the service, it would refuse to do so.

> I think a complementary PR would be to implement command-line arguments to `libertem-server` which allow pre-loading a cluster and a dataset / working directory (#1419). I will see if I can make this happen.

:+1:

> One method I can think of would be to support an `action=cluster` or similar in the URL path, then adjust the auto-launch URL from the python CLI to fill these parameters.

I think for pre-specifying compute resources, it would be easier to implement the whole procedure on the server side:

https://github.com/LiberTEM/LiberTEM/blob/3343014f58ff4f703eb5b471b0df559c74254c93/src/libertem/web/server.py#L213-L218

We already prime some settings into the `SharedState` instance there, it would be possible to add something like `shared_state.executor_state.set_executor(get_executor_from_args(args))` (pseudo-code, also need to handle params etc.). With the only hitch being that `set_executor` is currently `async`, but I think that can be fixed or worked around (by running it in the event loop).

**Comment by sk1p** [comment]
Currently investigating the CI failures, once they are resolved this should be good to go.

* Failure on Mac OS / Py3.9: seems to be very slow I/O on the CI worker, should be "resolved" on the next run (or we have to twiddle with the timeouts in CI)
* nodejs failure: need to revert the `package-lock.json` changes - the versions chosen in that update are not compatible
* `web_api` failure: currently looks like an issue related to the new Python 3.11.6 release, but is a bit hard to reproduce (conda doesn't have that version yet, so currently trying the `python:3.11.6-slim-bookworm` docker image)

**Comment by sk1p** [comment]
> * `web_api` failure: currently looks like an issue related to the new Python 3.11.6 release, but is a bit hard to reproduce (conda doesn't have that version yet, so currently trying the `python:3.11.6-slim-bookworm` docker image)

Cannot reproduce with  `python:3.11.6-slim-bookworm` , will work on the other issues first.

**Comment by sk1p** [comment]
> * Failure on Mac OS / Py3.9: seems to be very slow I/O on the CI worker, should be "resolved" on the next run (or we have to twiddle with the timeouts in CI)

That seems to be the case, no failure on this run.

> * nodejs failure: need to revert the `package-lock.json` changes - the versions chosen in that update are not compatible

Reverting the change and the lint run passes.

> * `web_api` failure: currently looks like an issue related to the new Python 3.11.6 release, but is a bit hard to reproduce (conda doesn't have that version yet, so currently trying the `python:3.11.6-slim-bookworm` docker image)

The reason it didn't reproduce was that I didn't rebase on top of master, which has the progress bar changes. Now, I can reproduce locally and work on a fix.

**Review by uellue — APPROVED** [review]
Discussion with @sk1p: Documentation can live as a separate item "LiberTEM as a service" in `docs/source/deployment/`.

Minus documentation and rebuilt client: Tested successfully, LGTM and many thanks! 👍 This will be great when deployed with the data management.

**Comment by sk1p** [comment]
> Discussion with @sk1p: Documentation can live as a separate item "LiberTEM as a service" in `docs/source/deployment/`.

This is now done. I'll push a client rebuild and things should be ready to be merged.

**Pull request merged**